frontend/src/home/graph-d3.js

```diff
@@ -0,0 +1,427 @@
+        let node = unseen.values().next().value;
+        unseen.delete(node);
+
+        // DFS for all reachable nodes, start with all directly reachable
+        let stack = [node];
+        while (stack.length > 0) {
+            // take top node on stack
+            let cur_node = stack.pop();
+
+            // visit it, give it a group
+            visited.add(cur_node);
+            group_map[cur_node] = group_id;
+
+            // add all directly reachable (and unvisited nodes) to stack
+            let directly_reachable = link_map[cur_node].filter(n => unseen.has(n));
+            stack = stack.concat(directly_reachable);
+
+            // remove all directly reachable elements from unseen
+            directly_reachable.forEach(n => unseen.delete(n));
+        }
+
+        group_id++;
+    }
+
+    return group_map;
+}
+
+function get_real_width(elem, props) {
+    // if we're given a static width, use it
+    if (props.width !== undefined) {
+        return props.width;
+    }
+
+    // calculate width of provided (likely, parent) element
+    let style = getComputedStyle(elem);
+    return elem.clientWidth - parseFloat(style.paddingLeft) - parseFloat(style.paddingRight);
+}
+
+
+// Graph -- store in an object to maintain state
+var d3Graph = {
+    SVG: null,
+    SLIDER: null,
+    G_NODE: null,
+    G_LINK: null,
+    NODE_DATA: null,
+    LINK_DATA: null,
+    COLOR: null,
+    DRAG_TARGET: null,
+    SIMULATION: null
+};
+
+
+// When there are exactly two nodes in the graph, d3 sets all kinds of things to
+// NaN for an unknown reason
+// This bool exists to mark all the places in the code where we implement the hack to fix this issue.
+d3Graph.HORRIBLE_TWO_NODE_HACK = true;
+
+
+// Initialize the data
+d3Graph.init_data = function(el, props, state) {
+    if (this.HORRIBLE_TWO_NODE_HACK) {
+        /// When there are exactly two nodes, add an additional one with an id of "" and an edge with value -1
+        if (state.graph.nodes.length === 2) {
+            state.graph.nodes.push({id: ""});
+            state.graph.links.push({source: state.graph.nodes[0], target: "", value: -1});
+            state.graph.data[""] = {is_archive: false};
+        }
+    }
+
+    this.NODE_DATA = state.graph.nodes;
+    this.LINK_DATA = state.graph.links;
+
+    // simulation
+    this.SIMULATION = d3.forceSimulation()
+        .force("link", d3.forceLink().id(d => d.id))
+        .force("charge", d3.forceManyBody().strength(-200).distanceMax(50).distanceMin(10))
+        .force("collision", d3.forceCollide().radius(d => props.radius * 2));
+
+    this.SIMULATION
+        .nodes(this.NODE_DATA);
+
+    this.SIMULATION.force("link")
+        .links(this.LINK_DATA);
+
+    // assign groups to nodes
+    let group_map = get_group_map(state.graph.links.map(d => ({source: d.source.id, target: d.target.id})));
+    state.graph.nodes.forEach(d => d.group = group_map[d.id]);
+
+    // set COLOR (function from group => color)
```

Comment:
Looks like this might be broken, there should be an equal number of colors as there are groups. (two in the example)

Reply:
Will look into this

Reply:
I think this (and the jumpy node problem) are due to two of my test files having been in the same directory, so there were two nodes with the same ID (which is based on directory path). I assume this is a limitation in compare50 I simply hadn't been aware of when generating the test data.

Reply:
@Jelleas it was definitely the test data, here's with different source directories: 
![image](https://user-images.githubusercontent.com/43184986/88840204-1f6b6980-d1aa-11ea-8f2e-337ed050605d.png)